### PR TITLE
Rust: Include parameters and patterns in the CFG

### PR DESCRIPTION
**Pull request by hvitved**
This PR adds parameters and patterns to the CFG, needed for SSA. Unlike expressions, which construct values, patterns destruct values, so they are modeled pre-order in the CFG. The first exception being [identifier patterns](https://doc.rust-lang.org/reference/patterns.html#identifier-patterns)

```rust
match value {
   x @ Option::None =>  // ..
   // ...
}
```
where we need to first match against `Option::None` and only bind `x` in case of a match, and the second exception being `|` patterns, which are modeled like logical disjunction expressions.

### Simple parameters

```rust
fn mutate_param(x : &mut i64) { // x
    *x = // $ read_access=x
        *x + // $ read_access=x
        *x; // $ read_access=x
}
```

<details><summary>Before</summary>

```mermaid
flowchart TD
1["enter mutate_param"]
10["... + ..."]
11["x"]
12["* ..."]
13["x"]
2["exit mutate_param"]
3["exit mutate_param (normal)"]
4["BlockExpr"]
5["* ..."]
6["... = ..."]
7["ExprStmt"]
8["x"]
9["* ..."]

1 --> 7
3 --> 2
4 --> 3
5 --> 11
6 --> 4
7 --> 8
8 --> 5
9 --> 13
10 --> 6
11 --> 9
12 --> 10
13 --> 12
```

</details>

<details><summary>After</summary>

```mermaid
flowchart TD
1["enter mutate_param"]
10["x"]
11["* ..."]
12["... + ..."]
13["x"]
14["* ..."]
15["x"]
2["exit mutate_param"]
3["exit mutate_param (normal)"]
4["x"]
5["Param"]
6["BlockExpr"]
7["* ..."]
8["... = ..."]
9["ExprStmt"]

1 --> 4
3 --> 2
4 -- match --> 5
5 --> 9
6 --> 3
7 --> 13
8 --> 6
9 --> 10
10 --> 7
11 --> 15
12 --> 8
13 --> 11
14 --> 12
15 --> 14
```

</details>

### Match expression

```rust
fn match_pattern5() {
    let either = Either::Left(32); // either
    match either { // $ read_access=either
        Either::Left(a3) | Either::Right(a3) // a3
            => print_i64(a3), // $ read_access=a3
    }
}
```

<details><summary>Before</summary>

```mermaid
flowchart TD
1["enter match_pattern5"]
10["MatchExpr"]
11["either"]
12["OrPat"]
13["PathExpr"]
14["CallExpr"]
15["a3"]
2["exit match_pattern5"]
3["exit match_pattern5 (normal)"]
4["BlockExpr"]
5["LetStmt"]
6["either"]
7["PathExpr"]
8["CallExpr"]
9["32"]

1 --> 5
3 --> 2
4 --> 3
5 --> 7
6 -- match, no-match --> 11
7 --> 9
8 --> 6
9 --> 8
10 --> 4
11 --> 12
12 -- match --> 13
13 --> 15
14 --> 10
15 --> 14
```

</details>

<details><summary>After</summary>

```mermaid
flowchart TD
1["enter match_pattern5"]
10["MatchExpr"]
11["either"]
12["TupleStructPat"]
13["[match(true)] OrPat"]
14["a3"]
15["TupleStructPat"]
16["a3"]
17["PathExpr"]
18["CallExpr"]
19["a3"]
2["exit match_pattern5"]
3["exit match_pattern5 (normal)"]
4["BlockExpr"]
5["LetStmt"]
6["either"]
7["PathExpr"]
8["CallExpr"]
9["32"]

1 --> 5
3 --> 2
4 --> 3
5 --> 7
6 -- match --> 11
7 --> 9
8 --> 6
9 --> 8
10 --> 4
11 --> 12
12 -- match --> 14
12 -- no-match --> 15
13 -- match --> 17
14 -- match --> 13
15 -- match --> 16
16 -- match --> 13
17 --> 19
18 --> 10
19 --> 18
```

</details>

### Parameter pattern matching

```rust
fn param_pattern2(
    (Either::Left(a9) | Either::Right(a9)): Either // a9
) -> () {
    print_i64(a9); // $ read_access=a9
}
```

<details><summary>Before</summary>

```mermaid
flowchart TD
1["enter param_pattern2"]
2["exit param_pattern2"]
3["exit param_pattern2 (normal)"]
4["BlockExpr"]
5["PathExpr"]
6["CallExpr"]
7["ExprStmt"]
8["a9"]

1 --> 7
3 --> 2
4 --> 3
5 --> 8
6 --> 4
7 --> 5
8 --> 6
```

</details>

<details><summary>After</summary>

```mermaid
flowchart TD
1["enter param_pattern2"]
10["BlockExpr"]
11["PathExpr"]
12["CallExpr"]
13["ExprStmt"]
14["a9"]
2["exit param_pattern2"]
3["exit param_pattern2 (normal)"]
4["Param"]
5["TupleStructPat"]
6["[match(true)] OrPat"]
7["a9"]
8["TupleStructPat"]
9["a9"]

1 --> 5
3 --> 2
4 --> 13
5 -- match --> 7
5 -- no-match --> 8
6 -- match --> 4
7 -- match --> 6
8 -- match --> 9
9 -- match --> 6
10 --> 3
11 --> 14
12 --> 10
13 --> 11
14 --> 12
```

</details>

### CODE REVIEW

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll [hunk 1]
```diff
@@ -141,11 +138,10 @@ class LogicalAndTree extends PostOrderTree, LogicalAndExpr {
 
 class BlockExprTree extends StandardPostOrderTree, BlockExpr {
   override AstNode getChildNode(int i) {
-    result = super.getStmtList().getStatement(i)
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:144]: Why this change? I see inconsistent use of `instanceof SomeClass` and `extends ..., SomeClass` in this file. Wouldn't it be better to consistently use `instanceof` ?

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll [hunk 2]
```diff
@@ -183,7 +179,26 @@ class CastExprTree extends StandardPostOrderTree instanceof CastExpr {
   override AstNode getChildNode(int i) { i = 0 and result = super.getExpr() }
 }
 
-class ClosureExprTree extends LeafTree instanceof ClosureExpr { }
+// Closures have their own CFG scope, so we need to make sure that their
+// CFG is not mixed with the surrounding CFG. This is done by retrofitting
+// `first`, `propagatesAbnormal`, and `succ` below.
+class ClosureExprTree extends StandardPostOrderTree, ClosureExpr {
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:185]: I'm not sure I like the way this is formulated. If I understand correctly you want the `succ` relation of a `StandardOrderTree` and the `first` and `last` predicates of  a `LeafTree`. Perhaps it would be clearer to make this class extend `StandardOrderTree` and avoid the `not succ = this` restriction in `succ` and add an explicit `last` predicate. It's about the same amount of lines and you avoid the tweaks to "undo" the behaviour of a `Post` order tree.

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll [hunk 3]
```diff
@@ -532,3 +569,164 @@ class YieldExprTree extends StandardPostOrderTree instanceof YieldExpr {
 class YeetExprTree extends StandardPostOrderTree instanceof YeetExpr {
   override AstNode getChildNode(int i) { i = 0 and result = super.getExpr() }
 }
+
+/**
+ * Provides `ControlFlowTree`s for patterns.
+ *
+ * Since patterns destruct values, they are modeled in pre-order, except for
+ * `OrPat`s and `IdentPat`s.
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:577]: Why are ident and or-patterns different?

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll [hunk 4]
```diff
@@ -532,3 +569,164 @@ class YieldExprTree extends StandardPostOrderTree instanceof YieldExpr {
...
+
+  class BoxPatTree extends PreOrderPatTree, BoxPat {
+    override Pat getPat(int i) { i = 0 and result = this.getPat() }
+  }
+
+  class RestPatTree extends LeafTree, RestPat { }
+
+  class LiteralPatTree extends LeafTree, LiteralPat { }
+
+  class MacroPatTree extends LeafTree, MacroPat { } // todo
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:674]: Done, in another PR ;-)

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll [hunk 5]
```diff
@@ -532,3 +569,164 @@ class YieldExprTree extends StandardPostOrderTree instanceof YieldExpr {
...
+
+  class TuplePatTree extends PreOrderPatTree, TuplePat {
+    override Pat getPat(int i) { result = this.getField(i) }
+  }
+
+  class TupleStructPatTree extends PreOrderPatTree, TupleStructPat {
+    override Pat getPat(int i) { result = this.getField(i) }
+  }
+
+  class ConstBlockPatTree extends LeafTree, ConstBlockPat { } // todo?
```
Comment (aibaars) [inline rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:731]: I think the constant value inside is evaluated compile-time, so perhaps having this as a `LeafTree` is the right thing to do.

File: rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll [hunk 6]
```diff
@@ -117,13 +117,61 @@ class BooleanCompletion extends ConditionalCompletion, TBooleanCompletion {
...
+    // `let` statements without an `else` branch must be exhaustive
+    pat = any(LetStmt let | not let.hasLetElse()).getPat()
+    or
+    // `match` expressions must be exhaustive, so last arm cannot fail
+    pat = any(MatchExpr me).getLastArm().getPat()
+    or
+    // parameter patterns must be exhaustive
+    pat = any(Param p).getPat()
+  ) and
+  not pat = any(ForExpr for).getPat() // workaround until `for` loops are desugared
```
[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll:140]: Thats because of how we currently model `for` loops; the pattern is used to represent the condition.

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll (hunk 1 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:144]: Better to consistently use `extends`.

File: rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll (hunk 3 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:577]: Because in `e @ pat` we only "evalute" `e` when `pat` matches. And in `pat1 | pat2` we need it to be post-order in order to get the correct dominance information needed for guards.

**Review by aibaars — APPROVED** [review]

File: rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll [hunk 7]
```diff
@@ -117,13 +117,61 @@ class BooleanCompletion extends ConditionalCompletion, TBooleanCompletion {
   override string toString() { result = "boolean(" + value + ")" }
 }
 
+/** Holds if node `pat` has the constant match value `value`. */
```
[the remark this answers is not in the document]
Comment (paldepind) [reply rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll]: I don't think this is correct use of "irrefutable". Irrefutable is property of a pattern, but this predicate is a property of a pattern as an AST node and considers the surrounding AST context as well.

For instance, in
```rust
match b { Some(a) => 1, None => 0 }
```
this predicate holds for `None` (since the last pattern in a `match` arm is guaranteed to succeed due to exhaustiveness) but `None` is not an irrefutable pattern.

I think we should rename this predicate (not sure to what though). We could also extract a proper `isIrrefutablePattern` predicate from this predicate (but it will probably be a conservative approximation as refutability for custom types depends on the type definition. E.g. the pattern `Foo(a)` is irrefutable if `Foo` the only case in the `enum` and refutable otherwise).

[the remark this answers is not in the document]
Comment (aibaars) [reply rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll]: Yeah, you're right. Perhaps `isExhaustiveMatch` or similar. An irrefutable pattern is exhaustive on its own, and a last pattern in `match` should be exhaustive.

[the remark this answers is not in the document]
Comment (paldepind) [reply rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll]: That sounds like a great name to me. I'll do a PR :)

[the remark this answers is not in the document]
Comment (paldepind) [reply rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll]: Done, PR is here https://github.com/github/codeql/pull/17751
